truss/tests/test_model_inference.py

```diff
@@ -938,3 +873,239 @@ def enable_gpu_fn(conf):
         # But make sure traces have parents at all.
         assert len(user_parents) > 3
         assert len(truss_parents) > 3
+
+
+# Returning Response Objects ###########################################################
+
+
+@pytest.mark.integration
+def test_truss_with_response():
+    """Test that user-code can set a custom status code."""
+    model = """
+    from fastapi.responses import Response
+
+    class Model:
+        def predict(self, inputs):
+            return Response(status_code=inputs["code"])
+    """
+    from fastapi import status
+
+    config = "model_name: custom-status-code-truss"
+
+    with ensure_kill_all(), temp_truss(model, config) as tr:
+        _ = tr.docker_run(local_port=8090, detach=True, wait_for_server_ready=True)
+
+        response = requests.post(PREDICT_URL, json={"code": status.HTTP_204_NO_CONTENT})
+        assert response.status_code == 204
+        assert "x-baseten-error-source" not in response.headers
+        assert "x-baseten-error-code" not in response.headers
+
+        response = requests.post(
+            PREDICT_URL, json={"code": status.HTTP_500_INTERNAL_SERVER_ERROR}
+        )
+        assert response.status_code == 500
+        assert response.headers["x-baseten-error-source"] == "04"
+        assert response.headers["x-baseten-error-code"] == "700"
+
+
+@pytest.mark.integration
+def test_truss_with_streaming_response():
+    # TODO: one issue with this is that (unlike our "builtin" streaming), this keeps
+    #  the semaphore claimed potentially longer if the client drops.
```

Comment:
good call -- although maybe now that you have access to the request, users can propagate a cancellation?

Reply:
Something like:

```
def predict(self, request: Request):
    def text_generator():
         if request.is_disconnected():
             return
             
         for i in range(3):
                yield f"data: {i}\\n\\n"
    return StreamingResponse(text_generator(), media_type="text/event-stream")    
```

Reply:
Cool, I actually wanted to add examples/test cases from the branch I made for cancellation over here.

Reply:
Done.